source/core/slang-byte-encode-util.cpp

```diff
@@ -0,0 +1,283 @@
+#include "slang-byte-encode-util.h"
+
+
+
+namespace Slang {
+
+// Descriptions of algorithms here...
+// https://github.com/stoklund/varint
+
+#if SLANG_LITTLE_ENDIAN && SLANG_UNALIGNED_ACCESS
+// Testing on i7, unaligned access is around 40% faster
+#   define SLANG_BYTE_ENCODE_USE_UNALIGNED_ACCESS 1
+#endif
+
+#ifndef SLANG_BYTE_ENCODE_USE_UNALIGNED_ACCESS
+#   define SLANG_BYTE_ENCODE_USE_UNALIGNED_ACCESS 0
+#endif
+
+#define SLANG_REPEAT_2(n) n, n
+#define SLANG_REPEAT_4(n) SLANG_REPEAT_2(n), SLANG_REPEAT_2(n)
+#define SLANG_REPEAT_8(n) SLANG_REPEAT_4(n), SLANG_REPEAT_4(n)
+#define SLANG_REPEAT_16(n) SLANG_REPEAT_8(n), SLANG_REPEAT_8(n)
+#define SLANG_REPEAT_32(n) SLANG_REPEAT_16(n), SLANG_REPEAT_16(n)
+#define SLANG_REPEAT_64(n) SLANG_REPEAT_32(n), SLANG_REPEAT_32(n)
+#define SLANG_REPEAT_128(n) SLANG_REPEAT_64(n), SLANG_REPEAT_64(n)
+
+/* static */const int8_t ByteEncodeUtil::s_msb8[256] =
+{
+    - 1, 
+    0, 
+    SLANG_REPEAT_2(1), 
+    SLANG_REPEAT_4(2), 
+    SLANG_REPEAT_8(3), 
+    SLANG_REPEAT_16(4), 
+    SLANG_REPEAT_32(5),
+    SLANG_REPEAT_64(6),
+    SLANG_REPEAT_128(7),
+};
+
+/* static */size_t ByteEncodeUtil::calcEncodeLiteSizeUInt32(const uint32_t* in, size_t num)
+{
+    size_t totalNumEncodeBytes = 0;
+    
+    for (size_t i = 0; i < num; i++)
+    {
+        const uint32_t v = in[i];
+
+        if (v < kLiteCut1)
+        {
+            totalNumEncodeBytes += 1;
+        }
+        else if (v <= kLiteCut1 + 255 * (kLiteCut2 - 1 - kLiteCut1))
+        {
+            totalNumEncodeBytes += 2;
+        }
+        else
+        {
+            totalNumEncodeBytes += calcNonZeroMsByte32(v) + 2;
+        }
+    }
+    return totalNumEncodeBytes;
+}
+
+/* static */size_t ByteEncodeUtil::encodeLiteUInt32(const uint32_t* in, size_t num, uint8_t* encodeOut)
+{
+    uint8_t* encodeStart = encodeOut;
+
+    for (size_t i = 0; i < num; ++i)
+    {
+        uint32_t v = in[i];
+
+        if(v < kLiteCut1)
+        {
+            *encodeOut++ = uint8_t(v);
+        }
+        else if (v <= kLiteCut1 + 255 * (kLiteCut2 - 1 - kLiteCut1))
```

Comment:
So, is the motivation here to shift things more heavily toward 1- and 2-byte encodings? Compared to a pure "prefix varint" encoding, this seems to complicate the decoding logic by adding more branches (you need to test against the two pre-defined cutoff points). Do the commit messages have some discussion of the variations you tried here?

This isn't important to finalize right now, since we can continue to tweak the encoding as long as we don't make the binary format public/stable.